backend/modules/parsers/multi_modal_parser.py

```diff
@@ -87,129 +82,89 @@ async def call_vlm_agent(
-                                    final_texts.append(
-                                        Document(
-                                            page_content="File Name: "
-                                            + file_name
-                                            + "\n\n"
-                                            + page_content,
-                                            metadata={
-                                                "image_b64": pages[pg_no],
-                                                "page_number": pg_no,
-                                                "source": file_name,
-                                            },
-                                        )
-                                    )
-                else:
-                    logger.debug("No response from VLM...")
-                await asyncio.sleep(5)
-            return final_texts
-        except Exception as e:
-            logger.exception(f"Final Exception: {e}")
-            raise e
+            await asyncio.sleep(5)
+
+        return final_texts
+
+    def _get_pages(self, filepath: str) -> Dict[int, str]:
+        if filepath.endswith(MULTI_MODAL_PARSER_SUPPORTED_PDF_EXTENSION):
+            return self._get_pdf_pages(filepath)
+        elif filepath.endswith(MULTI_MODAL_PARSER_SUPPORTED_IMAGE_EXTENSIONS):
+            return self._get_image_page(filepath)
+        else:
+            raise ValueError(
+                "Invalid file extension. Supported formats: PDF, PNG, JPEG, JPG"
+            )
+
+    def _get_pdf_pages(self, filepath: str) -> Dict[int, str]:
+        pages = {}
+        doc = fitz.open(filepath)
+        for page in doc:
+            try:
+                pix = page.get_pixmap(alpha=False)
+                img = np.frombuffer(pix.samples, dtype=np.uint8).reshape(
+                    (pix.h, pix.w, pix.n)
+                )
+                _, buffer = cv2.imencode(".png", img)
+                pages[page.number + 1] = base64.b64encode(buffer).decode("utf-8")
+            except Exception as e:
+                logger.exception(f"Error in page {page.number + 1}: {e}")
+        return pages
+
+    def _get_image_page(self, filepath: str) -> Dict[int, str]:
+        with open(filepath, "rb") as f:
+            return {0: base64.b64encode(f.read()).decode("utf-8")}
+
+    def _chunk_pages(self, data: Dict[int, str], size: int = 30):
+        it = iter(data)
+        for i in range(0, len(data), size):
+            yield {k: data[k] for k in islice(it, size)}
+
+    def _create_document(
+        self, file_name: str, pg_no: int, page_content: str, image_b64: str
+    ) -> Document:
+        return Document(
+            page_content=f"File Name: {file_name}\n\n{page_content}",
+            metadata={
+                "image_b64": image_b64,
+                "page_number": pg_no,
+                "source": file_name,
+            },
+        )
```

Comment:
I see that pre-signed uri is added post parsing stage. when chunks are returned. Just a thought, would it be beneficial to add the signed uri here it self, even if we don't, there is one missing step, we are not updating the chunk metadata with document metadata `_metadata` arg from `get_chunks` function. 

Please check for other parsers as well.

Reply:
When I implemented this step, I was getting a signed uri on demand using the internal fqn for the data point, the frontend preview code was designed that way too was wondering if we should change that. Moving forward additional data sources may not live on our own storage providers either right? for example external airtable or SQL dbs etc.

Reply:
Generating a pre-signed read URL in case of a tfy data dir required tfy fqn, which is not available in this function and will require many changes. Also, this will involve changing all the parsers. I don't think that's a good idea.

Reply:
@mnvsk97 - I suppose the chunk metadata already has an fqn, you might want to checked the indexed chunks within the qdrant that should help generate pre-signed urls in the query controller step itself. 

As now I see, there is a big flaw here being, the pre-signed url generated in indexing stage becomes part of chunk metadata embedded in the qdrant, as a result signed-url does not change. Hence, after sometime, the pre-signed url will become invalid.

A good way is to - 
- Remove signed url generation code from the indexing and parsing stage
- Add a `_generate_pre_signed_urls` function to `BaseQueryController`, this function is called at the last step of querycontroller when `answer` and `docs` are generated, that uses doc_fqn to generate the url
- Send this pre-signed url as `url` key in the metadata for each chunk when returning the answers
- If you see fqn, is not present, we can add the code for that in data-loading stage and update the chunk metadata with document metadata in the parsing stage.

Reply:
Nice catch. I'll move this to `QueryController`

Reply:
Resolved in latest commits